### PR TITLE
bpo-46640: Py_NAN now uses the C99 NAN constant

### PR DESCRIPTION
**Pull request by vstinner**
Building Python now requires a C99 <math.h> header file providing the
NAN constant.

https://bugs.python.org/issue46640

### CODE REVIEW

**Review by corona10 — APPROVED** [review]
lgtm

**Comment by vstinner** [comment]
test_urllib2 failed on many buildbots: https://bugs.python.org/issue36019

**Comment by corona10** [comment]
> test_urllib2 failed on many buildbots

That's true :(

**Review by raghavthind2005 — APPROVED** [review]

File: Doc/whatsnew/3.11.rst
Comment (mdickinson) [inline Doc/whatsnew/3.11.rst]: As mentioned on the issue, this is a big deal - by requiring the `NAN` constant, we'd be effectively requiring IEEE 754 floating-point, which is something that CPython has been reluctant to do to date.

I'm +1 on requiring IEEE 754 for CPython, but that should be an explicit decision that has buy-in from the rest of the core developers. So I think this needs discussion on python-dev before we can make this change.

Note that the other requirements listed here are not at all similar: those are all functions that are _required_ to be present by C99, while the ``NAN`` constant may or may not exist according to the standard.  (In the words of the standard, it's "defined if and only if the implementation supports quiet NaNs for the float type".)

File: Include/pymath.h
Comment (mdickinson) [inline Include/pymath.h]: Here you're changing the type of the `Py_NAN` constant from `double` to `float`. That _might_ be benign, or it might not. Either way, it's probably worth using `((double)NAN)` instead of plain `NAN` as the macro value.

Comment (vstinner) [reply Include/pymath.h]: Would it make sense to declare a Py_NAN constant? Something like: ``PyAPI_DATA(double) Py_NAN;``.

The macro NAN can require computing the value or call a function. I don't think that <math.h> ensures that it's directly a value, even if it's the case if most cases.

File: Doc/whatsnew/3.11.rst
Comment (mdickinson) [reply Doc/whatsnew/3.11.rst]: Nitpick: `isinf`, `isnan` and `isfinite` are macros, not functions.

File: Include/pymath.h
Comment (mdickinson) [reply Include/pymath.h]: > Would it make sense to declare a Py_NAN constant?

If we can make that work, then sure. How would the constant value be assigned?

File: Doc/whatsnew/3.11.rst
Comment (mdickinson) [reply Doc/whatsnew/3.11.rst]: BTW, should this change have a separate news entry, given that the changes from issue bpo-45440 are already released?

Comment (mdickinson) [reply Doc/whatsnew/3.11.rst]: Gah. Sorry, this is what's new, not Misc/NEWS. Ignore me.

File: Misc/NEWS.d/next/Build/2022-02-04-21-26-50.bpo-46640.HXUmQp.rst
Comment (mdickinson) [inline Misc/NEWS.d/next/Build/2022-02-04-21-26-50.bpo-46640.HXUmQp.rst]: Why are there two separate `Misc/NEWS.d` files for this PR?

Comment (mdickinson) [reply Misc/NEWS.d/next/Build/2022-02-04-21-26-50.bpo-46640.HXUmQp.rst]: Ah, I see. One's for the build changes, one for the C API changes. Got it, but the wording here needs updating, since it's only true that building requires the `NAN` constant on a platform that doesn't define `Py_NO_NAN`.

Comment (vstinner) [reply Misc/NEWS.d/next/Build/2022-02-04-21-26-50.bpo-46640.HXUmQp.rst]: I removed it.

File: Include/pymath.h
Comment (vstinner) [reply Include/pymath.h]: > It's probably worth using ((double)NAN) instead

Oh, I didn't notice that it's a float! Fixed.

File: Doc/whatsnew/3.11.rst
Comment (vstinner) [reply Doc/whatsnew/3.11.rst]: It remains possible to build Python without Py_NAN: define Py_NO_NAN in pyconfig.h. I clarified that in the change documentation.

**Comment by vstinner** [comment]
I updated my PR:

* Add a single Build NEWS entry -remove the C API entry).
* Clarify that it remains possible to build Python without NaN support: define Py_NO_NAN in pyconfig.h.
* Use ``__builtin_nan("")`` if available (GCC, clang).

**Comment by vstinner** [comment]
> Nitpick: isinf, isnan and isfinite are macros, not functions.

For Python, a macro or function is the same. This change with this NEWS entry is already released: was moved to ``NEWS.d/3.11.0a2.rst``. I prefer to leave it unchanged. Or do you think that it matters?

**Comment by vstinner** [comment]
I looked at the machine code generated by ``clang -O3`` on Fedora 35 x86-64 (clang 13.0.0). For this mathmodule.c function:

```
static double
m_atan2(double y, double x)
{
    if (Py_IS_NAN(x) || Py_IS_NAN(y))
        return Py_NAN;
  ...
}
```

The ``return Py_NAN;`` becomes:

```
=> 0x00007fffea51a2a2 <m_atan2+98>:	movsd  xmm0,QWORD PTR [rip+0x1e1e]        # 0x7fffea51c0c8
   0x00007fffea51a2aa <m_atan2+106>:	ret
```

The `movsd` instruction sets ``xmm0`` to:

```
(gdb) p $xmm0
$1 = {
  v8_bfloat16 = {0, 0, 0, nan(0x78), 0, 0, 0, 0},
  v4_float = {0, nan(0x780000), 0, 0},
  v2_double = {nan(0x8000000000000), 0},
  v16_int8 = {0, 0, 0, 0, 0, 0, -8, 127, 0, 0, 0, 0, 0, 0, 0, 0},
  v8_int16 = {0, 0, 0, 32760, 0, 0, 0, 0},
  v4_int32 = {0, 2146959360, 0, 0},
  v2_int64 = {9221120237041090560, 0},
  uint128 = 9221120237041090560
}
```

It seems like clang creates a private variable at ``0x7fffea51c0c8`` with ``v2_double = {nan(0x8000000000000), 0}``, to at the end, it's a single 128-bit MOV instruction to write directly to the SSE XMM0 register.

**Comment by vstinner** [comment]
``gcc -03`` (GCC 11.2.1) produces similar code for ``return Py_NAN;``:

```
0x00007fffea5177d8 <math_atan2+728>:   movsd  xmm0,QWORD PTR [rip+0x7790]   # 0x7fffea51ef70
```

**Comment by vstinner** [comment]
Using ``gcc -O3``, ``#define Py_NAN (__builtin_nan(""))`` and ``#define Py_NAN ((double)NAN)`` produce the same machine code and use the same number (``nan(0x8000000000000)``).

**Comment by vstinner** [comment]
The Windows SDK defines these macros in `ucrt/corecrt_math.h`:

```
#ifndef _HUGE_ENUF
    #define _HUGE_ENUF  1e+300  // _HUGE_ENUF*_HUGE_ENUF must overflow
#endif

#define INFINITY   ((float)(_HUGE_ENUF * _HUGE_ENUF))
#define NAN        ((float)(INFINITY * 0.0F))
```

ref: https://github.com/tpn/winsdk-10/blob/master/Include/10.0.16299.0/ucrt/corecrt_math.h

It's similar to Python current definition of Py_NAN (simplified code):

```
#define Py_HUGE_VAL HUGE_VAL
#define Py_NAN (Py_HUGE_VAL * 0.)
```

``Py_HUGE_VAL`` and ``Py_NAN`` are double.

File: Doc/whatsnew/3.11.rst
Comment (mdickinson) [inline Doc/whatsnew/3.11.rst:610]: It turns out that building under `Py_NO_NAN` doesn't actually work right now. I've opened an issue: https://bugs.python.org/issue46656

File: Include/pymath.h
Comment (mdickinson) [inline Include/pymath.h]: With this change, the comment about `INF*0` or `INF/INF` working is out of date; that sentence should probably be removed.

**Review by mdickinson — APPROVED** [review]
Needs a comment update, but otherwise LGTM. It turns out that Py_NO_NAN doesn't actually work right now, but that's a separate issue.

**Comment by vstinner** [comment]
Merged. Thanks for the review @mdickinson!

**Pull request merged**
